# Complexiteit van sorteren
Er zijn diverse soorten sorteer-algoritmen. 
In deze opgave gaan we een aantal van deze algoritmen op hun run time efficiëntie vergelijken; 
eerst door het draaien van een paar tests, vervolgens door analyse van de code (complexiteitsanalyse).
Bron: https://canvas.hu.nl/courses/7560/assignments/69747 

In [1]:
from numpy import random
import pandas as pd
from typing import List
import sys
import timeit
sys.path.append('/Users/wolfsinem/High-Performanced-Programming/functions/')

from sort_functions import selection_sort, recursive_selection_sort, insertion_sort, recursive_insertion_sort, merge_sort, recursive_merge_sort, merge_arrays, recursive_insertion
# %load_ext line_profiler #load line profiler extension
# %lprun -f selection_sort selection_sort(lijst_1)

#### Random lijsten genereren om de tests op uit te voeren

In [2]:
def random_list(num_steps):
    return list(random.sample(num_steps))

list1 = random_list(1000)
list2 = random_list(10000)
list3 = random_list(30000)

lists = [list1,list2,list3]

In [3]:
"""function that can be used for each algorithm to measure the execution time"""
def set_time(function,list_):
    for i in list_:
        start_time = timeit.default_timer()
        function(i)
        print('This algorithm took {:.4f} seconds'.format((timeit.default_timer() - start_time)))

In [4]:
# sorted list
sorted_list = [[i for i in range(1000)]]

# reversed list
reversed_list = sorted_list[::-1]

### Selection sort

In [5]:
set_time(selection_sort,lists)

This algorithm took 0.0403 seconds
This algorithm took 3.8713 seconds
This algorithm took 34.4327 seconds


- selection sort with sorted list

In [6]:
set_time(selection_sort,sorted_list)

This algorithm took 0.0367 seconds


- selection sort with reversed list 

In [7]:
set_time(selection_sort,reversed_list)

This algorithm took 0.0386 seconds


### Insertion sort

In [8]:
set_time(insertion_sort,lists)

This algorithm took 0.0003 seconds
This algorithm took 0.0028 seconds
This algorithm took 0.0097 seconds


- insertion sort with sorted list

In [9]:
set_time(insertion_sort,sorted_list)

This algorithm took 0.0003 seconds


- insertion sort with reversed list

In [10]:
set_time(insertion_sort,reversed_list)

This algorithm took 0.0004 seconds


### Merge sort

In [11]:
set_time(merge_sort,lists)

This algorithm took 0.0026 seconds
This algorithm took 0.0175 seconds
This algorithm took 0.0437 seconds


- merge sort with sorted list

In [12]:
set_time(insertion_sort,sorted_list)

This algorithm took 0.0001 seconds


- merge sort with reversed list

In [13]:
set_time(insertion_sort,reversed_list)

This algorithm took 0.0003 seconds


### Tabel met alle uitkomsten

| list | selection_sort | insertion_sort | merge_sort |
|------:|----------------|----------------|------------|
| 1000  | 0.0403              | 0.0003              | 0.0026          |
| 10000 | 3.8713              | 0.0028              | 0.0175          |
| 30000 | 34.4327              | 0.0097              | 0.0437          |



| list | selection_sort | insertion_sort | merge_sort |
|------:|----------------|----------------|------------|
| sorted  | 0.0367              | 0.0003              | 0.0001          |
| reversed | 0.0386              | 0.0004              | 0.0003          |


## Theorie
<i> Bekijk / bepaal aan de hand van de algoritmes (en beschrijvingen) hierboven, wat de theoretische run time efficiëntie (Big O) van elk van deze algoritmes is. Bepaal hiervoor 'best case', 'worst case' en 'average case' run time efficiëntie. </i>

### selection_sort
- best case: O(n^2)
- worst case: O(n^2) 
- average case: O(n^2)


### insertion_sort
- best case: O(n)
- worst case: O(n^2)
- average case: O(n^2)


### merge_sort
- best case: O(nlog(n))
- worst case: O(nlog(n)) 
- average case: O(nlog(n))

<b>Maakt het voor de complexiteit (Big O) van een algoritme uit of je een iteratieve of een recursieve versie beschouwt?</b>

Hoewel er een verschil zit in gebruik van een iteratieve en recursieve methode maakt het voor de Big O complexiteit niet uit welke methode je beschouwt. De tijd complexiteit blijft enigszins hetzelfde maar zal het verschil meer liggen in 'space complexity'.

- Stel je maakt gebruik van een algoritme, dan zal deze dezelfde volgordelijke stappen ondernemen voor zowel een iteratieve als recursieve methode.